### PR TITLE
Enable index config by default.

### PR DESCRIPTION
**Pull request by gortiz**
This is a bugfix.

As introduced in #10183, each `IndexConfig` subclass had to manually check for `enabled` nullability in its own way. It is correct that some indexes may request to explicitly ask for `enabled: true` in order to be enabled, but all default indexes should be enabled if their config object is present. 

For example:

```js
{
   bloom: {
       fpp: 0.02,
   }
}
```

Should be considered as a enabled config. Even the empty object should be considered enabled

```js
{
   bloom: {
   }
}
```

In order to explicitly disable an index one must write

```js
{
   bloom: {
       enabled: "false"
   }
}
```

If the index key is not present or set to null, then the default value will be used. In that case, whether the index is enable or not depends on the index type. For example forward index is enabled by default.

Previous code was erroneously initializing some index configs as disabled, as detected in #10554 for bloom filter. Affected indexes were:
- Bloom
- Dictionary
- H3
- FST
- Range
- Text
- Inverted

This PR adds a constructor in `IndexConfig` that contains the default logic, which is the one all default indexes should use. The PR also add unit tests that verify the correct deserialization of indexes for all index types.

### CODE REVIEW

**Comment by jadami10** [comment]
Thank you for the awesome test coverage. I'll let others review as well. In case there's some concern with always enabling an index unless the config is exactly `"enabled": false`, I definitely think the way you did it here is the least surprising for users.

File: pinot-spi/src/main/java/org/apache/pinot/spi/config/table/IndexConfig.java
Comment (KKcorps) [inline pinot-spi/src/main/java/org/apache/pinot/spi/config/table/IndexConfig.java:47]: Do we need this constructor? We can directly set the value it in the previous one right?

Comment (gortiz) [reply pinot-spi/src/main/java/org/apache/pinot/spi/config/table/IndexConfig.java:47]: I usually try to keep older constructors to do not break binary backward compatibility, but I don't think anyone had time to add code that relay on this constructor. Another reason to keep it here may be to skip boxing/unboxing, but this shouldn't be in the hotpath.

So I'm going to remove it.

**Review by Jackie-Jiang — APPROVED** [review]
I still think treating `null` boolean as `true` is counter intuitive, and can lead to surprises similar to this one. E.g. `Boolean.parseBoolean()` will treat everything other than `"true"` as `false`, including `null`. I'd suggest changing the field to be `disabled` which by default is false to avoid this confusion.
LGTM otherwise

**Comment by gortiz** [comment]
I still think from a user perspective is cleaner to say

```js
{
  bloom: {
    enabled: false
  }
}
```

Than
```js
{
  bloom: {
    disabled: true
  }
}
```

But I'm open to change that if other people think like @Jackie-Jiang. What I think that is important is to decide that ASAP. Once people start using `enabled` it would be more difficult to change it.

**Comment by jadami10** [comment]
> One less common example that IMO can lead to surprising behavior is:
> 
> ```
> {
>   enabled: null
> }
> ```
> 
> The above example will actually enable the index

It's surprising, but I think it's still better. If you're configuring an index in any way, you probably already want it on. And if you don't want it on, you're either going to delete the entire block, or explicitly set `"enabled": false`.

As for enabled vs disabled, was this a new a field, or did it already exist? Because if it existed, wouldn't this be a big breaking change now?

**Comment by Jackie-Jiang** [comment]
> As for enabled vs disabled, was this a new a field, or did it already exist? Because if it existed, wouldn't this be a big breaking change now?

It is newly introduced, and we are debating whether we should explicitly put `enabled: false` or `disabled: true` to disable an index. I'm suggesting using `disabled: true` to disable the index, where `disabled: null` will be treated as index enabled

**Comment by gortiz** [comment]
I think we can discuss about that, but it shouldn't block this PR. We can merge this to solve the problem and then decide whether we want to change the attribute from `enabled` to `disabled`

**Comment by gortiz** [comment]
> What happens if we deploy this, then it switches? I guess enabled: false is the only case where the behavior will switch?

That is my concern. The easiest change would be to just ignore the `enabled` attribute, so:

| when enabled was  | prev this PR  | after this PR  | if we use disabled   |
|---|---|---|---|
| not defined  |  sometimes disabled the index | enable the index  | enable the index |
| null  | sometimes disable the index  | enable the index  | enable the index  |
| true  | enable the index  | enable the index  | enable the index  |
| false  | disable the index  | disable the index  | **enable the index**  |

**Comment by gortiz** [comment]
Let's move the discussion to https://github.com/apache/pinot/pull/10585

**Pull request merged**
